ni_measurementlink_service/session_management/_reservation.py

```diff
@@ -183,6 +275,61 @@ def _create_sessions_core(
                 )
             yield typed_session_infos
 
+    def _get_connection_core(
+        self,
+        session_type: Type[TSession],
+        pin_or_relay_names: Union[str, Iterable[str], None] = None,
+        sites: Union[int, Iterable[int], None] = None,
+        instrument_type_id: Optional[str] = None,
+    ) -> TypedConnection[TSession]:
+        results = self._get_connections_core(
+            session_type, pin_or_relay_names, sites, instrument_type_id
+        )
+        if len(results) > 1:
+            raise ValueError(
+                "Too many reserved connections matched the specified criteria. "
+                f"Expected single connection, got {len(results)} connections."
+            )
+        return results[0]
+
+    def _get_connections_core(
+        self,
+        session_type: Type[TSession],
+        pin_or_relay_names: Union[str, Iterable[str], None] = None,
+        sites: Union[int, Iterable[int], None] = None,
+        instrument_type_id: Optional[str] = None,
+    ) -> Sequence[TypedConnection[TSession]]:
+        pin_order = _to_iterable(pin_or_relay_names, self._reserved_pin_or_relay_names)
+
+        if sites == SITE_ALL_SITES:
+            site_order = self._reserved_sites
+        else:
+            site_order = _to_iterable(sites, self._reserved_sites)
+            if SITE_SYSTEM_PINS in self._reserved_sites and SITE_SYSTEM_PINS not in site_order:
```

Comment:
@pbirkhol-ni @jasonmreding @jonathanmendez Does this approach to handling system pins make sense to you? If any of the reserved pins are system pins and the specified list of sites does not include -1, add -1 to the end.

Reply:
LabVIEW does something similar. However, I just always add -1 to the end of the sites list (unless -1 is already included). I suppose excluding -1 if there are no reserved system pins is reasonable, but it requires tracking extra state that I don't currently track.

Reply:
> I suppose excluding -1 if there are no reserved system pins is reasonable, but it requires tracking extra state that I don't currently track.

The pin map context's site list doesn't include site -1, so this didn't work correctly anyway.

Reply:
This code now adds -1 to the end, if not already present.

Reply:
FYI, the pin map service will return an invalid argument error if you try and pass a pin site of -1. It will return it as a site number in the output returned from the API but not as an input to the API.